### ISSUE TITLE
Localhost image linking issue? - (renamed from Custom script overlay instance?)

### ISSUE THREAD

**Issue by awrebels** [Firebottle/Firebot#425]
#### Describe your issue or feature request (details are helpful!):

With the regards to the overlay instances, what is the correct effects response structure for a custom script?

**Comment by awrebels** [comment]
I figure out that in the json file it is "overlayInstance" however I can't quite get it to work with a custom script. Wondering if you have similar problems or likely it is just me. :)

**Comment by awrebels** [comment]
ok i think i diagnosed the issue and apologies for the incorrect title. I am pulling via api an external image and this is the error (below) I get when I use the overlay with localhost:XXXX. However if i use the old .html overlay, the image does in fact show up. Perhaps i'm missing something with my set up?

{"status":"error","message":"/resource/jvna2wulg1eqhjs1370hpvic3ki2yl3qudqq7feyl79qw7b9?time=1510460806000 not found"}

**Comment by awrebels** [comment]
Is there an html version of each of the overlay instances or are they only localhost?

**Comment by awrebels** [comment]
ok i retract that last question. using local file plus ?instance=XYZ works just fine.

**Comment by Firebottle** [comment]
Sorry for the delay. Were you able to fix the issue?

**Comment by awrebels** [comment]
I couldn't quite figure out the localhost solution. Maybe it is some cross-domain issue. I did eventually realize the html file location + ?instance=XYZ worked but still confused about the localhost issue.

**Comment by ebiggz** [comment]
So just to clarify, you are doing an image effect or html effect via custom script and you are trying to assign it to a overlay instance?

**Comment by awrebels** [comment]
I'm doing an html effect via custom script, grabbing an image through an api, and assigning it to a custom overlay instance, using localhost which does not show the image. If I instead use the html instance (with no other adjustments) the image via html does show up.

**Comment by ebiggz** [comment]
hmm. The error youre getting makes it seem like you are using an image effect, not an HTML effect. I am having trouble replicating this issue. I have custom script written as so:

```js
exports.run = function(runRequest) {
	return new Promise(resolve => {
		resolve({
			success: true,
			effects: [
				{
					type: "HTML",
					enterAnimation: "fadeIn",
					exitAnimation: "fadeOut",
					html: `<img class="img-test" src="https://4fi8v2446i0sw2rpq2a3fg51-wpengine.netdna-ssl.com/wp-content/uploads/2016/06/KittenProgression-Darling-week7.jpg">`,
					length: "5",
					removal: "img-test",
					overlayInstance: "test"
				}
			]
		});
	});
}
```

And I am properly seeing the image show up in the "test" overlay instance. aka: `http://localhost:7473/overlay/?instance=test`

**Comment by awrebels** [comment]
I'm sorry you are absolutely correct. I was looking at the wrong code. I am using an image effect, which I see only has local file support. Although it technically works using html overlay

**Comment by ebiggz** [comment]
No worries! 

Accessing local files is a bit more tricky now that we switched over to the hosted overlay. 
Built in security measures prevent html from accessing local files (and for good reason). 

This causes an issue, though, for effects that try to access local files (ie image, video, html effects)
So we had to build an endpoint (the "/resource/" one you see in the error) that requests a file from the http server. We use automatically generated tokens to request files as a layer of security so that the endpoint couldnt be used to request any old file on the pc. This token stuff happens behind the scenes for image and video effects but currently there is no way for custom scripts to generate a token to use for local files used in a raw html effect. 

I need to look into allowing custom scripts to do this, just need to make sure there isn't any major security ramifications in allowing that.

Also, I've always thought the image effect should have support for online images instead of only local, since the video effect basically has that with the youtube option. I'll create a feature request for that so we have it (if there isnt one already, I cant remember offhand). 

Hopefully those two changes will resolve the issues you are having and make the system more flexible.
